ktor-server/ktor-server-plugins/ktor-server-forwarded-header/jvmAndNix/src/io/ktor/server/plugins/forwardedheaders/XForwardedHeaders.kt

```diff
@@ -0,0 +1,206 @@
+    ) {
+        val protoValues = headers.protoHeader?.split(',')
+        val httpsFlagValues = headers.httpsFlagHeader?.split(',')
+        val hostValues = headers.hostHeader?.split(',')
+        val portValues = headers.portHeader?.split(',')
+        val forValues = headers.forHeader?.split(',')
+
+        protoValues?.let { values ->
+            val scheme = extractValue(values) ?: return@let
+            connectionPoint.scheme = scheme
+            URLProtocol.byName[scheme]?.let { connectionPoint.port = it.defaultPort }
+        }
+
+        httpsFlagValues?.let { values ->
+            val useHttps = extractValue(values).toBoolean()
+            if (!useHttps) return@let
+
+            connectionPoint.let { route ->
+                route.scheme = "https"
+                route.port = URLProtocol.HTTPS.defaultPort
+            }
+        }
+
+        hostValues?.let { values ->
+            val hostAndPort = extractValue(values) ?: return@let
+            val host = hostAndPort.substringBefore(':')
+            val port = hostAndPort.substringAfter(':', "")
+
+            connectionPoint.host = host
+            port.toIntOrNull()?.let {
+                connectionPoint.port = it
+            } ?: URLProtocol.byName[connectionPoint.scheme]?.let {
+                connectionPoint.port = it.defaultPort
+            }
+        }
+
+        portValues?.let { values ->
+            val port = extractValue(values) ?: return@let
+            connectionPoint.port = port.toInt()
+        }
+
+        forValues?.let { values ->
+            val remoteHost = extractValue(values) ?: return@let
+            if (remoteHost.isNotBlank()) {
+                connectionPoint.remoteHost = remoteHost
+            }
+        }
+    }
+
+    private fun String?.toBoolean() = this == "yes" || this == "true" || this == "on"
+
+}
+
+/**
+ * Values of X-Forward-* headers. Each property may contain multiple comma-separated values.
+ */
+public data class XForwardedHeaderValues(
+    /**
+     * Comma-separated list of values for [Config.protoHeaders] header
+     */
+    public val protoHeader: String?,
+    /**
+     * Comma-separated list of values for [Config.forHeaders] header
+     */
+    public val forHeader: String?,
+    /**
+     * Comma-separated list of values for [Config.hostHeaders] header
+     */
+    public val hostHeader: String?,
+    /**
+     * Comma-separated list of values for [Config.httpsFlagHeaders] header
+     */
+    public val httpsFlagHeader: String?,
+    /**
+     * Comma-separated list of values for [Config.portHeaders] header
+     */
+    public val portHeader: String?
+)
+
+/**
+ * [XForwardedHeaders] plugin allows you to obtain information headers from the original request in reverse proxy
+ * setups. For more information see https://datatracker.ietf.org/doc/html/rfc7239
+ */
+public val XForwardedHeaders: ApplicationPlugin<Application, XForwardedHeadersConfig, PluginInstance> =
```

Comment:
I would also keep `XForwardedHeaderSupport` without implementation, but with `@Deprecated` and `replaceWith`